src/verifier.rs

```diff
@@ -0,0 +1,72 @@
+use crate::{Error, Keystore, Message, PublicKey, SecretKey, Signature};
+
+use borrown::Borrown;
+
+/// Signature verifier based on a keystore
+pub trait Verifier {
+    /// Keystore error implementation
+    type Error: From<Error> + From<<Self::Keystore as Keystore>::Error>;
+
+    /// Concrete keystore implementation
+    type Keystore: Keystore;
+
+    /// Accessor to the keystore
+    ///
+    /// Might fail if the keystore is in a corrupt state, not initialized or locked by a
+    /// concurrent thread.
+    fn keystore(&self) -> Result<&Self::Keystore, Self::Error>;
+
+    /// Secret key indexed by `id`.
+    fn id_secret(
+        &self,
+        id: &<Self::Keystore as Keystore>::KeyId,
+    ) -> Result<Borrown<'_, SecretKey>, Self::Error> {
+        let keystore = self.keystore()?;
+        let secret = keystore.secret(id)?.ok_or(Error::KeyNotFound)?;
+
+        Ok(secret)
+    }
+
+    /// Public key indexed by `id`.
+    fn id_public(
+        &self,
+        id: &<Self::Keystore as Keystore>::KeyId,
+    ) -> Result<Borrown<'_, PublicKey>, Self::Error> {
+        let keystore = self.keystore()?;
+        let public = keystore.public(id)?.ok_or(Error::KeyNotFound)?;
+
+        Ok(public)
+    }
+
+    /// Verify a given message with the public key identified by `id`
+    fn verify(
+        &self,
+        id: &<Self::Keystore as Keystore>::KeyId,
```

Comment:
It seems like this is the only reason for coupling signature verification with the keystore. I'm having a hard time imagining what this would be used for though.

Reply:
The advantage of making verification stateless and not relying on a keystore is that it would be much easier to port into no-std environments.

Reply:
The idea is that the validity for the consensus is not restricted to the validity of the signature itself, but also its existence in a given repository of public keys (e.g. validators).

We could indeed just store the plain key instead of the ID, but here we are using a common pattern of the keystore

Reply:
Typically a keystore is only used for storing the private keys. It seems like a strange abstraction to me to use a "keystore" for public keys only in an ACL context.

Reply:
Its just an indexed and persisted set of keys. In our case, its indexed by a round range. The sole purpose of this was to simplify the implementation and rely on a structure that already exists (keystore mapping an arbitrary index to a key, not necessarily a keypair). However, it can be also implemented ad-hoc as part of the consensus.

Reply:
`Verifier` removed in favor of validators set that will supersede its functionality: https://github.com/FuelLabs/fuel-crypto/pull/12/commits/b82e8577cf13d97180d0a8ba9f38db37cbe77b37